File: backend/app/domain/layout_switch.py
"""布局切换的兼容性判定与槽位重映射。

只开放能完整安置当前页所有块的布局；不引入「未放置」块，也不丢弃内容。
匹配结果必须确定可重复，因此按视觉阅读顺序固定候选顺序后再回溯。
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol

from app.domain.content import BlockType
from app.domain.layout import Layout, Slot, get_layout, load_layouts

_TYPE_LABELS: dict[str, str] = {
    "text": "文字",
    "bullets": "要点",
    "image": "图片",
    "chart": "图表",
    "table": "表格",
    "kpi": "指标",
}


class BlockLike(Protocol):
    id: str
    type: BlockType | str
    slot_id: str


@dataclass(frozen=True)
class LayoutSwitchOk:
    mapping: dict[str, str]


@dataclass(frozen=True)
class LayoutSwitchErr:
    reason: str


LayoutSwitchResult = LayoutSwitchOk | LayoutSwitchErr


@dataclass(frozen=True)
class LayoutCandidate:
    layout_id: str
    name: str
    usage: str
    compatible: bool
    reason: str | None
    current: bool


def _as_block(block: BlockLike | dict[str, Any]) -> tuple[str, str, str]:
    if isinstance(block, dict):
        return str(block["id"]), str(block["type"]), str(block["slot_id"])
    return str(block.id), str(block.type), str(block.slot_id)


def _slot_order_key(slot: Slot) -> tuple[float, float, str]:
    return (slot.rect.y, slot.rect.x, slot.id)


def _sort_slots(slots: list[Slot]) -> list[Slot]:
    return sorted(slots, key=_slot_order_key)


def _sort_blocks(
    blocks: list[tuple[str, str, str]],
    current: Layout,
) -> list[tuple[str, str, str]]:
    """按块在原布局中的视觉位置排序；未知槽位排到最后以保持确定。"""

    def key(item: tuple[str, str, str]) -> tuple[float, float, str]:
        _block_id, _block_type, slot_id = item
        slot = current.slot_by_id(slot_id)
        if slot is None:
            return (1e9, 1e9, slot_id)
        return (slot.rect.y, slot.rect.x, slot.id)

    return sorted(blocks, key=key)


def _type_label(block_type: str) -> str:
    return _TYPE_LABELS.get(block_type, block_type)


def _explain_incompatible(
    blocks: list[tuple[str, str, str]],
    target: Layout,
) -> str:
    # 优先给出类型级原因，便于界面说明「为什么置灰」
    type_counts: dict[str, int] = {}
    for _block_id, block_type, _slot_id in blocks:
        type_counts[block_type] = type_counts.get(block_type, 0) + 1

    for block_type, count in sorted(type_counts.items()):
        capacity = sum(1 for slot in target.slots if block_type in slot.accepts)
        if count > capacity:
            label = _type_label(block_type)
            return f"目标布局最多放 {capacity} 个{label}块，当前页有 {count} 个"

    present = set(type_counts)
    required = [slot for slot in target.slots if slot.required]
    for slot in _sort_slots(required):
        if not any(accepted in present for accepted in slot.accepts):
            labels = "、".join(_type_label(item) for item in slot.accepts)
            return f"目标布局需要{labels}内容，当前页没有"

    if len(blocks) > len(target.slots):
        return f"目标布局只有 {len(target.slots)} 个槽位，当前页有 {len(blocks)} 个内容块"

    if len(blocks) < len(required):
        return f"目标布局需要至少 {len(required)} 个内容块，当前页只有 {len(blocks)} 个"

    return "当前页内容无法完整安置到目标布局的槽位"


def _backtrack(
    blocks: list[tuple[str, str, str]],
    slots: list[Slot],
    used: set[str],
    mapping: dict[str, str],
) -> dict[str, str] | None:
    if len(mapping) == len(blocks):
        required_ok = all((not slot.required) or slot.id in mapping.values() for slot in slots)
        return mapping if required_ok else None

    block_id, block_type, _ = blocks[len(mapping)]
    for slot in slots:
        if slot.id in used:
            continue
        if block_type not in slot.accepts:
            continue
        used.add(slot.id)
        mapping[block_id] = slot.id
        found = _backtrack(blocks, slots, used, mapping)
        if found is not None:
            return found
        del mapping[block_id]
        used.remove(slot.id)
    return None


def plan_layout_switch(
    blocks: list[BlockLike] | list[dict[str, Any]],
    current_layout_id: str,
    target_layout_id: str,
) -> LayoutSwitchResult:
    """判定目标布局是否兼容，并给出块 id → 新 slot_id 的完整映射。"""
    current = get_layout(current_layout_id)
    target = get_layout(target_layout_id)
    normalized = [_as_block(block) for block in blocks]

    # 已在当前布局上：身份映射即可，避免空页/残缺页把「当前」标成不可选
    if target_layout_id == current_layout_id:
        return LayoutSwitchOk({block_id: slot_id for block_id, _type, slot_id in normalized})

    ordered_blocks = _sort_blocks(normalized, current)
    ordered_slots = _sort_slots(list(target.slots))
    mapping = _backtrack(ordered_blocks, ordered_slots, set(), {})
    if mapping is None:
        return LayoutSwitchErr(_explain_incompatible(normalized, target))
    return LayoutSwitchOk(mapping)


def list_layout_candidates(
    blocks: list[BlockLike] | list[dict[str, Any]],
    current_layout_id: str,
) -> list[LayoutCandidate]:
    """返回全部布局的候选列表，供接口直接驱动置灰。"""
    layouts = load_layouts()
    candidates: list[LayoutCandidate] = []
    for layout_id in sorted(layouts):
        layout = layouts[layout_id]
        result = plan_layout_switch(blocks, current_layout_id, layout_id)
        compatible = isinstance(result, LayoutSwitchOk)
        candidates.append(
            LayoutCandidate(
                layout_id=layout.id,
                name=layout.name,
                usage=layout.usage,
                compatible=compatible,
                reason=None if compatible else result.reason,
                current=layout_id == current_layout_id,
            )
        )
    return candidates

File: java_backend/src/main/java/com/aippt/domain/LayoutSwitch.java
package com.aippt.domain;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.aippt.domain.content.Blocks;

public final class LayoutSwitch {

    private static final Map<String, String> TYPE_LABELS = Map.of(
            "text", "文字",
            "bullets", "要点",
            "image", "图片",
            "chart", "图表",
            "table", "表格",
            "kpi", "指标",
            "cards", "卡片",
            "callout", "提示"
    );

    public record Ok(Map<String, String> mapping) {
    }

    public record Err(String reason) {
    }

    public record Candidate(
            String layoutId,
            String name,
            String usage,
            boolean compatible,
            String reason,
            boolean current
    ) {
    }

    private LayoutSwitch() {
    }

    public static Object plan(List<Map<String, Object>> blocks, Layout current, Layout target) {
        List<String[]> normalized = asBlocks(blocks);
        if (current.id().equals(target.id())) {
            Map<String, String> mapping = new LinkedHashMap<>();
            for (String[] item : normalized) {
                mapping.put(item[0], item[2]);
            }
            return new Ok(mapping);
        }
        List<String[]> orderedBlocks = sortBlocks(normalized, current);
        List<Layout.Slot> orderedSlots = sortSlots(target.slots() == null ? List.of() : target.slots());
        Map<String, String> mapping = backtrack(orderedBlocks, orderedSlots, new HashSet<>(), new LinkedHashMap<>());
        if (mapping == null) {
            return new Err(explain(normalized, target));
        }
        return new Ok(mapping);
    }

    public static List<Candidate> candidates(List<Map<String, Object>> blocks, Layout current, Map<String, Layout> layouts) {
        List<Candidate> result = new ArrayList<>();
        for (String layoutId : layouts.keySet().stream().sorted().toList()) {
            Layout layout = layouts.get(layoutId);
            Object planned = plan(blocks, current, layout);
            boolean compatible = planned instanceof Ok;
            result.add(new Candidate(
                    layout.id(),
                    layout.name(),
                    layout.usage(),
                    compatible,
                    compatible ? null : ((Err) planned).reason(),
                    layoutId.equals(current.id())
            ));
        }
        return result;
    }

    private static List<String[]> asBlocks(List<Map<String, Object>> blocks) {
        List<String[]> result = new ArrayList<>();
        for (Map<String, Object> block : blocks) {
            result.add(new String[]{Blocks.id(block), Blocks.type(block), Blocks.slotId(block)});
        }
        return result;
    }

    private static List<Layout.Slot> sortSlots(List<Layout.Slot> slots) {
        return slots.stream()
                .sorted((a, b) -> {
                    int byY = Double.compare(a.rect().y(), b.rect().y());
                    if (byY != 0) {
                        return byY;
                    }
                    int byX = Double.compare(a.rect().x(), b.rect().x());
                    return byX != 0 ? byX : a.id().compareTo(b.id());
                })
                .toList();
    }

    private static List<String[]> sortBlocks(List<String[]> blocks, Layout current) {
        List<String[]> copy = new ArrayList<>(blocks);
        copy.sort((a, b) -> {
            Layout.Slot left = current.slotById(a[2]);
            Layout.Slot right = current.slotById(b[2]);
            double ly = left == null ? 1e9 : left.rect().y();
            double ry = right == null ? 1e9 : right.rect().y();
            int byY = Double.compare(ly, ry);
            if (byY != 0) {
                return byY;
            }
            double lx = left == null ? 1e9 : left.rect().x();
            double rx = right == null ? 1e9 : right.rect().x();
            int byX = Double.compare(lx, rx);
            if (byX != 0) {
                return byX;
            }
            String lid = left == null ? a[2] : left.id();
            String rid = right == null ? b[2] : right.id();
            return lid.compareTo(rid);
        });
        return copy;
    }

    private static Map<String, String> backtrack(
            List<String[]> blocks,
            List<Layout.Slot> slots,
            Set<String> used,
            Map<String, String> mapping
    ) {
        if (mapping.size() == blocks.size()) {
            boolean requiredOk = slots.stream().allMatch(slot ->
                    !Boolean.TRUE.equals(slot.required()) || mapping.containsValue(slot.id())
            );
            return requiredOk ? mapping : null;
        }
        String[] block = blocks.get(mapping.size());
        for (Layout.Slot slot : slots) {
            if (used.contains(slot.id()) || slot.accepts() == null || !slot.accepts().contains(block[1])) {
                continue;
            }
            used.add(slot.id());
            mapping.put(block[0], slot.id());
            Map<String, String> found = backtrack(blocks, slots, used, mapping);
            if (found != null) {
                return found;
            }
            mapping.remove(block[0]);
            used.remove(slot.id());
        }
        return null;
    }

    private static String explain(List<String[]> blocks, Layout target) {
        Map<String, Integer> typeCounts = new HashMap<>();
        for (String[] block : blocks) {
            typeCounts.merge(block[1], 1, Integer::sum);
        }
        for (String type : typeCounts.keySet().stream().sorted().toList()) {
            int capacity = 0;
            if (target.slots() != null) {
                for (Layout.Slot slot : target.slots()) {
                    if (slot.accepts() != null && slot.accepts().contains(type)) {
                        capacity++;
                    }
                }
            }
            if (typeCounts.get(type) > capacity) {
                return "目标布局最多放 " + capacity + " 个" + TYPE_LABELS.getOrDefault(type, type)
                        + "块，当前页有 " + typeCounts.get(type) + " 个";
            }
        }
        Set<String> present = typeCounts.keySet();
        List<Layout.Slot> required = target.slots() == null ? List.of() : target.slots().stream()
                .filter(slot -> Boolean.TRUE.equals(slot.required()))
                .toList();
        for (Layout.Slot slot : sortSlots(required)) {
            if (slot.accepts() == null || slot.accepts().stream().noneMatch(present::contains)) {
                String labels = String.join("、", slot.accepts().stream()
                        .map(item -> TYPE_LABELS.getOrDefault(item, item))
                        .toList());
                return "目标布局需要" + labels + "内容，当前页没有";
            }
        }
        int slotCount = target.slots() == null ? 0 : target.slots().size();
        if (blocks.size() > slotCount) {
            return "目标布局只有 " + slotCount + " 个槽位，当前页有 " + blocks.size() + " 个内容块";
        }
        if (blocks.size() < required.size()) {
            return "目标布局需要至少 " + required.size() + " 个内容块，当前页只有 " + blocks.size() + " 个";
        }
        return "当前页内容无法完整安置到目标布局的槽位";
    }
}
